flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java

```diff
@@ -213,6 +217,22 @@ public void processElement(StreamRecord<RowData> element) throws Exception {
     });
   }
 
+  /**
+   * If upsert is enabled, turn update_after to insert if there isn't update_after followed by update_before.
+   * But it may lead to some unnecessary delete data if there are some data interspersed with other primary key data.
+   * e.g. K1-UB, K2-UB, K1-UA, K2-UB
+   */
+  private void processMultiUpdateAfter(RowData row) {
```

Comment:
Maybe we should also consider this scenario:
Key1-UB, Key1-UA, Key2-UB, Key1-UA.

Reply:
I will track ua before in keys set to fix this case.